modules/frontend/frontend.go

```diff
@@ -55,19 +56,19 @@ func New(cfg Config, next http.RoundTripper, o overrides.Interface, reader tempo
 		return nil, fmt.Errorf("query backend after should be less than or equal to query ingester until")
 	}
 
-	retryWare := newRetryWare(cfg.MaxRetries, registerer)
-
-	// cache
-	searchCache := newFrontendCache(cacheProvider, cache.RoleFrontendSearch, logger)
+	retryWare := pipeline.NewRetryWare(cfg.MaxRetries, registerer)
+	cacheWare := pipeline.NewCachingWare(cacheProvider, cache.RoleFrontendSearch, logger)
+	statusCodeWare := pipeline.NewStatusCodeAdjustWare()
 
 	// inject multi-tenant middleware in multi-tenant routes
 	traceByIDMiddleware := MergeMiddlewares(
 		newMultiTenantMiddleware(cfg, combiner.NewTraceByID, logger),
 		newTraceByIDMiddleware(cfg, o, logger), retryWare)
 
-	searchMiddleware := MergeMiddlewares(
-		newMultiTenantMiddleware(cfg, combiner.NewSearch, logger),
-		newSearchMiddleware(cfg, o, reader, searchCache, logger), retryWare)
+	searchPipeline := pipeline.Build(
+		asyncPipeline(cfg, newAsyncSearchSharder(reader, o, cfg.Search.Sharder, logger), logger),
```

Comment:
This kind of hides the multi-tenant handling.  Is there a way to "flatten" it up to this level like the other apis are built?

Reply:
good call. it's nice to see all the pipeline components in one spot. changed to:
```
	searchPipeline := pipeline.Build(
		[]pipeline.AsyncMiddleware[*http.Response]{
			multiTenantMiddleware(cfg, logger),
			newAsyncSearchSharder(reader, o, cfg.Search.Sharder, logger),
		},
		[]pipeline.Middleware{cacheWare, statusCodeWare, retryWare},
		next)
```